### DISCUSSION TITLE
Parent with children view

### DISCUSSION THREAD

**Discussion by mjashanks** [Budibase/budibase#3223] [Ideas]
Often we like to be able to display a detail screen, with a collection of child records below it.

For example `One Project has many Tasks` - the component tree looks like:

```
> Data Provider - Project by _id, LIMIT 1
  > Repeater
      > Form for project
      > Data Provider - Relationship Project.Tasks
          > New Task button (will create new task for this project, via a modal)
          > Table (or repeater)
```

We should develop a block component for this.

**Comment by gvlx** [comment]
Questions:
1) Why the the need for a repeater?
The first data provider should already supply all data, but in fact, without the repeater, there is nothing available..
2) The "New Task button" action should provide the id to link the new data back in to this relationship automatically.
3) The modal opened by the "New Task button" has to contain a "Close Screen Modal" action on its "Save" button.

**suggestion**
A modal dialog where it asks the user for an existing screen or to create a new screen to input the related data.

**Reply by gvlx** [reply]
You've already described much of the actions needed here: https://docs.budibase.com/guides/create-a-child-record-from-the-parent

**Reply by gvlx** [reply]
However the instructions for editing a child record from a parent require some 'extra magical steps'. 🪄
I'll post them another day. :alarm_clock:

**Reply by gvlx** [reply]
Here it is #3365 , complete with the 'extra magical steps'! 🪄

**Comment by neun3D** [comment]
Hello guys.
novice in Budibase community here (take that in consideration :))
is there a component in Budibase like TreeView 
![image](https://user-images.githubusercontent.com/22809478/174246962-516209b3-6e45-470f-abeb-7435e0569ef3.png)
that is implemented in Budibase editor. Or if not can I build it? Can you point me in which direction to go?
Thank you in advance!

**Comment by melohagan** [comment]
Hey @neun3D 

This guide may give you some ideas to get you started!

Assuming the following table relationship:
- One Information -> Many Services

![treeview](https://user-images.githubusercontent.com/101575380/174786516-16d548e6-f68c-4d7f-a9eb-ed890e396f06.gif)

### Step 1 - Components
Add a Data Provider for the *Information* table, alongside a Repeater. 
Inside the Repeater add a Container with *Row Direction* and give it a border:
<img width="258" alt="Screenshot 2022-06-21 at 12 05 41" src="https://user-images.githubusercontent.com/101575380/174785146-c5aee25b-38cf-41e3-85cf-56b0fb331f22.png">

Inside the Container, add a expand Button with the chevron character as its text `⌄`
Click *Edit custom CSS* to get the desired style, and add a little margin:
<img width="1225" alt="Screenshot 2022-06-21 at 12 09 27" src="https://user-images.githubusercontent.com/101575380/174785788-6f585fad-9231-4fc6-aeae-11660ec12d25.png">

Also add a Headline for the *Information* title:
![Screenshot 2022-06-21 at 12 10 51](https://user-images.githubusercontent.com/101575380/174786014-aa0ee668-969f-4418-b4e0-136aefa44fd4.png)

Finally add a Cards Block to the Repeater. Choose the Services relationship as the datasource:
![Screenshot 2022-06-21 at 12 15 03](https://user-images.githubusercontent.com/101575380/174786701-8e13c5f5-7829-483a-8243-5969b33540f5.png)

Final component tree:
![Screenshot 2022-06-21 at 12 17 51](https://user-images.githubusercontent.com/101575380/174787150-278b8930-299b-4881-94d1-634cfde200a5.png)
*I'll cover the Global Style a little later on*

### Step 2 - Keeping track of state
We need to keep track of which containers should be expanded or closed. This can be done through App State.

Click *Define actions* for the expand button. Add the *Update State* actions as follows:
![Screenshot 2022-06-21 at 12 20 46](https://user-images.githubusercontent.com/101575380/174787613-cbb16a17-0bfb-4e6c-bbcb-560dd6e32ec3.png)

JavaScript binding:
```
let expanded = $("State.Expanded");
if (!expanded) {
	expanded = {}
}
expanded[$("Repeater.01_information._id")] = !expanded[$("Repeater.01_information._id")];

return expanded;
```
This will store a map for each information row with a simple true/false boolean. True means expanded, false means closed.

### Step 3 - Conditional UI
Now we only want to show the *Service* Card Block for a given information if it should be expanded as determined by the app state. 

For the *Service* Card Block, click *Configure conditions* (at the bottom of the Settings panel).
Add a condition as follows:
![Screenshot 2022-06-21 at 12 26 30](https://user-images.githubusercontent.com/101575380/174788606-7d0b06b3-947b-49ae-881d-1fe53f06985d.png)

JavaScript binding:
`return $("State.Expanded")[$("Repeater.01_information._id")];`

### Step 4 (Optional) - Further expand button styling
To prevent the chevron expand button from showing a highlighted background colour on hover, we need to override the spectrum styling with an Embed component.

Add the Embed component to the top of your component tree. Add the following content:
```
<style>
.spectrum-Button--primary:hover {
  background-color: var(--spectrum-button-primary-m-background-color, var(--spectrum-alias-background-color-transparent)) !important;
  color: green !important;
}
</style>
``` 
The *color* property sets the highlight colour of the chevron.

**Comment by neun3D** [comment]
@melohagan man thank You very much. This is terrific jumpstart.

**Reply by melohagan** [reply]
You're welcome - happy building!
